[Feat] 2차 과제 구현

devxsby opened the pull request:
## 🧣 2주차 과제 작업 내용 요약

`iPhone SE (3rd generation)`

![Simulator Screen Recording - iPhone SE (3rd generation) - 2022-04-19 at 17 38 04](https://user-images.githubusercontent.com/80062632/163961369-d0522a59-2164-4c6a-8d93-efe60fe8f23b.gif)

- 오토레이아웃 설정
- 탭 바 컨트롤러 구성
- 화면 전환 추가

---

## 🧤 PR Point

1. 로그인 뷰에서 처음에 인스타그램 로고 Image View의 Top을 기준으로 생각하고 피그마에 나온 13mini 기준의 Constrant값을 주고 그 다음 Object들을 종속적으로 연결했습니다. <br> 그러다 보니 화면 사이즈가 작아지면 이미지 뷰가 너무 아래로 내려와서 밑에 Object들이 잘리기 직전까지 나오더라구요. <br> 화면 크기에 맞게 Constrant를 조정하자니 그럼 오토 레이아웃을 한 의미가 없고, 조정을 안하자니 보기가 예쁘지 않은데요. <br> 다들 어떻게 구성하셨는지 궁금해요. 

>  저는 오토레이아웃을 지정할 때 TextField나 Button, Image View 등에 `Aspect Ratio`를 지정해서 화면 사이즈가 변경될 때 반응해서 같이 조정되게 구현했는데요. 실제로 이 방법을 많이 쓰는지 궁금합니다. 그리고 Label의 `Autoshrink`값도 Minimum Font Scale을 조정했는데요. 굳이 안해도 되는 것 같긴 한데 이 기능도 많이 사용하는지 궁금합니다.

2. 심화 과제 구현을 AppDelegate에서 SceneDelegate를 수정하고 window의 rootViewController 설정하려했는데 keyWindow 등 자꾸 에러가 떠서 쉬운 코드로 변경하였습니다. 이 방식으로 구현하신 분 어떻게 코드 작성하셨는지 궁금해요 !!! <br> (추가로 이렇게 구현할 경우 MVC패턴이 망가질 수 있다는 글도 읽었는데 왜 그런가요?) 🤔

3. @pcsoyeon 님의 폴더링을 참고했습니다.  MVC 패턴은 적용할 수 있으면 다음에 할 수 있도록 노력해볼게요. <br> 다른 분들은 어떤 방식으로 폴더링 하셨는지 비법 전수 부탁드려요 ~~!

<img width="345" alt="스크린샷 2022-04-19 17 13 02" src="https://user-images.githubusercontent.com/80062632/163957460-ffee1c31-d717-47e0-a95f-bf9f636bbd27.png">
 

4. 이 외 칭 찬 및 피 드 백 환 영 💥

---

## 🧦 스크린샷

### 1. 기본 과제 (1) : 오토레이아웃 설정

`로그인 화면 예시`
| `iPhone 13 Pro` | `iPhone 13 mini` | `iPhone SE (3rd generation)` | 
| :---: | :---: | :---: |
|![Simulator Screen Shot - iPhone 13 Pro Max ](https://user-images.githubusercontent.com/80062632/163945192-74e218bd-09d6-41af-8739-424b3afd7782.png) | ![Simulator Screen Shot - iPhone 13 mini](https://user-images.githubusercontent.com/80062632/163945242-da7e12ad-e4a3-422a-bbfc-8c298ae1d51c.png) | ![Simulator Screen Shot - iPhone SE (3rd generation)](https://user-images.githubusercontent.com/80062632/163949029-14a5f983-4859-4155-bc6e-2ab50235856a.png) |

### 2. 기본 과제 (2) : 탭 바 컨트롤러 구성 

| `Home` | `Search` | `Reels` | `Shop` | `Profile` | 
|  :---: | :---: | :---: | :---: | :---: |
|![Simulator Screen Shot - iPhone 13 mini - 2022-04-19 at 16 43 15](https://user-images.githubusercontent.com/80062632/163952011-fc092932-241c-472e-a121-00687934c02a.png)|![Simulator Screen Shot - iPhone 13 mini - 2022-04-19 at 16 43 18](https://user-images.githubusercontent.com/80062632/163952080-73e06696-5fbe-4e10-abc8-263a92cdb785.png)|![Simulator Screen Shot - iPhone 13 mini - 2022-04-19 at 16 43 20](https://user-images.githubusercontent.com/80062632/163952127-45935c3c-181a-4c0a-bccc-390d23cb2ded.png)|![Simulator Screen Shot - iPhone 13 mini - 2022-04-19 at 16 43 22](https://user-images.githubusercontent.com/80062632/163952212-0c750044-c057-4770-ad27-8245f9de1b2a.png)| ![Simulator Screen Shot - iPhone 13 mini - 2022-04-19 at 16 43 24](https://user-images.githubusercontent.com/80062632/163952262-d0ff110f-b9f6-43da-a2b5-cce74aab05e2.png)|

### 3. 기본 과제 (3) : 화면 전환 추가하기

| `AuthCompleteView` | `LoginView` | `TabBarMain` |
| :---: | :---: | :---: |
|![Simulator Screen Shot - iPhone 13 mini - 2022-04-19 at 16 48 17](https://user-images.githubusercontent.com/80062632/163952897-592300d3-9380-46eb-8985-f165370c3ae3.png)|![Simulator Screen Shot - iPhone 13 mini - 2022-04-19 at 16 48 56](https://user-images.githubusercontent.com/80062632/163953171-0e7c6352-026e-4bb3-a459-090b0bdf4c2a.png)|![Simulator Screen Shot - iPhone 13 mini - 2022-04-19 at 16 49 37](https://user-images.githubusercontent.com/80062632/163952984-3dda4cab-86c7-49eb-a866-65752c0af084.png)|
| 기준 화면 | 완료하기 버튼 누르면 <br> 탭 바 메인 화면으로 이동 | 다른 계정으로 로그인 누르면 로그인 <br> 화면으로 돌아가고 텍스트 필드 지워져 있음 |

### 4. 도전 과제 : 릴스 탭 바 아이콘 클릭 시 탭 바 속성 변경

| `Home` | `Reels` |`Profile` | 
| :---: | :---: | :---: |
| ![Simulator Screen Shot - iPhone 13 mini - 2022-04-19 at 16 58 12](https://user-images.githubusercontent.com/80062632/163954477-9f8efb3f-a56b-454f-a8c8-f04dbd772865.png)| ![Simulator Screen Shot - iPhone 13 mini - 2022-04-19 at 16 58 18](https://user-images.githubusercontent.com/80062632/163954504-1cbb9049-d50d-4b17-8694-844db1dd5d06.png)|![Simulator Screen Shot - iPhone 13 mini - 2022-04-19 at 16 58 23](https://user-images.githubusercontent.com/80062632/163954607-c89abc33-acc8-4c99-9e82-93cfaf655ed6.png)|
| 비교 화면1 : 홈 탭 바 | 릴스 탭 바 클릭 시 <br> 탭 바 아이템, 에셋 색상 변경 | 비교 화면2 : 프로필 탭 바 |

### 5. 심화 과제 : 로그인, 회원가입 플로우 이후 메인 화면 전환 시 적절한 화면 전환 방식 생각해보기

- rootViewController 를 변경해주었습니다.

`HOW?`
1. AuthCompleteViewController에서 TabBarMainViewController를 present로 띄우는 코드를 삭제하였습니다.
2. 루트 뷰로 대체하고자 하는 뷰 컨트롤러인 TabBarMainViewController를 rootVC로 인스턴스화 하였습니다.
3. window?코드를 구현했습니다. 
> self.view.window?.rootViewController = rootVC <br> self.view.window?.makeKeyAndVisible()

[`window.makeKeyAndVisible()` 공식 문서](https://developer.apple.com/documentation/uikit/uiwindow/1621610-makekey) <br>
 - keyWindow로 설정합니다.
 - keyWindow: window가 여러개 존재할 때, 가장 앞쪽에 배치된 window를 `key window`라고 지칭합니다.
 - window의 rootViewController를 위에서 세팅해주고, makeKeyAndVisible() 부르면 지정한 rootViewController가 상호작용을 받는 현재 화면으로 세팅을 완료합니다.

---

## 🧢 관련 이슈

- Resolved: [Feat] iOS 2차 과제 #3

devxsby committed:
[Add] 2차 과제 에셋 추가 (#3)

.

devxsby committed:
[Update] Extension 파일 중 불필요한 코드 삭제 (#3)

Extension의 UIView와 UITextField 파일 중 사용하지 않는 코드를 삭제했습니다.

devxsby committed:
[Chore] 이미지 에셋 Render As Original image로 수정 (#3)

.

devxsby committed:
[Style] 이미지 에셋 icn_back.png 파일명 변경 (#3)

.

devxsby committed:
[Add] 컬러 에셋 추가 (#3)

시스템 전반에 사용하는 light_blue, sky_blue, dark_grey 컬러 추가

devxsby committed:
[Feat] 커스텀 컬러 에셋 사용 (#3)

RGB코드 대신 컬러 에셋에 있는 sky_blue와 light_blue를 사용하였습니다.

devxsby committed:
[Feat] 오토 레이아웃 적용 (#3)

피그마에 올라온 iPhone 13 mini 사이즈에 맞게 적용하였습니다.

devxsby committed:
[Style] 변수명 userName 수정 (#3)

뷰에서 사용자 이름 입력 messageName으로 넘어가던 값을 userName으로 직관적으로 알아볼 수 있도록 수정했습니다.

devxsby committed:
[Style] Extension의 UIColor 파일의 컬러 에셋 코드 변수명 수정 (#3)

light_blue, sky_blue, dark_grey 세 가지 컬러셋을 수정하였습니다.

devxsby committed:
[Feat] Extension UIViewController+의 BackButtonCustom 추가 (#3)

뒤로가기 버튼 커스텀 함수를 익스텐션에 넣었습니다.
이미지 에셋을 추가하고, 타이틀을 제거하였습니다.
LoginViewController와 SignUpNameViewController의 setUI함수에서 호출되도록 하였습니다.

devxsby committed:
[Feat] SignUpPasswordViewController에서 눈 아이콘 버튼 추가 (#3)

SignUpPasswordViewController에서 눈 아이콘 버튼을 클릭 시 비밀번호 보안을 풀고 확인할 수 있는 기능을 추가하였습니다.

devxsby committed:
[Chore] 텍스트 필드 유효값을 확인하는 코드 수정 (#3)

통일성을 위해 .text에서 .isEmpty를 확인 하는 기존 코드에서 .hasTex에서 .isEnabled 방식으로 변경하였습니다.

devxsby committed:
[Refactor] setUI 커스텀 함수 추가 (#3)

기존 코드에서 setUI 함수로 한 번에  커스텀 UI 세팅을 하는 함수를 추가하였습니다.
이외 추가 설명을 위한 주석들을 추가하였습니다.

devxsby committed:
[Update] Source 파일 중 불필요한 항목 삭제 (#3)

Source의 Screen에서 HomeModel, HomeView, HomeViewController를 삭제했습니다.

devxsby committed:
[Refactor] Auth 구현 스토리보드 뷰 분리 (#3)

Main 스토리보드에 모아져 있던 뷰들을  Login, SignUp, AuthComplete 스토리 보드로 분리했습니다.

devxsby committed:
[Add] TabBar 구현을 위한 파일 생성 (#3)

TabBar를 구성하는 스토리보드 5개와 메인 탭 바 스토리보드를 생성했습니다.
2차 과제 구현 확인을 위해 Info.plist의 시작 Storyboard Name을 TabBarMain으로 변경했습니다.

devxsby committed:
[Fix] 뒤로가기 아이콘 파일명 수정 (#3)

버그로 보이지 않았던 @2x, @3x 파일명을 수정했습니다.

devxsby committed:
[Fix] Tab Bar 아이콘 순서 변경 및 버그 수정 (#3)

탭 바 아이콘 순서를 home, search, reels, shop, profile로 정렬했습니다.
추가로 릴스 아이콘이 보이지 않던 버그를 수정했습니다.

devxsby committed:
[Feat] Auth Complete 뷰에서 화면 전환 추가 (#3)

기존 완료하기 버튼을 누르면 현재 뷰가 닫혔던 것에서 메인 탭 바 화면으로 전환시켰습니다.
다른 계정으로 로그인하기 버튼을 추가하여 로그인 화면으로 돌아가도록 했습니다.

devxsby committed:
[Feat] Login 뷰에서 이전에 작성한 계정의 TextField 지우는 기능 추가 (#3)

로그인 화면에서 로그인 버튼과 회원가입 버튼을 클릭 시 다음 뷰로 넘어가고, 이전에 작성한 계정의 nameTextField와 passwordTextField가 사라지도록 하였습니다.

devxsby committed:
[Update] 스토리보드 오토레이아웃 미세 조정 (#3)

.

devxsby committed:
[Chore] 생명주기 viewWillAppear 추가 (#3)

deleteAccount 함수를 .viewWillAppear()에서 구현하도록 수정했습니다.

devxsby committed:
[Chore] 이미지 에셋 Render As Default image로 수정 (#3)

릴스 탭 바 커스텀을 위해 이미지 에셋 Render As 값을 Default로 변경합니다.

devxsby committed:
[Feat] 도전과제 - 릴스 탭 바 클릭 시 아이콘 색상 반전 (#3)

가운데 릴스 탭 바 아이콘을 눌렀을 때 구현된 내용입니다.
1. 탭 바 아이템의 색상 변경
2. 추가한 에셋 색상 변경

devxsby committed:
[Docs] README 2주차 과제 테이블 체크 (#3)

devxsby committed:
[Chore] 불필요한 코드 삭제 (#3)

UITabBarController+ 파일에서 호출하는 코드가 중복 작성되어 삭제했습니다.

lsj8706 commented on Instagram-Clone/Instagram-Clone/Source/Screen/Auth/Controller/LoginViewController.swift:
```diff
@@ -56,27 +61,38 @@ class LoginViewController:
+        
+        /// TextField
+        nameTextField.clearButtonMode = .whileEditing
+        nameTextField.attributedPlaceholder = NSAttributedString(string: "전화번호, 사용자 이름 또는 이메일", attributes: [NSAttributedString.Key.foregroundColor : UIColor(named: "dark_grey")!])
```
NSAttributedString으로 placeholder 커스텀까지....!
꼼꼼함 👍

devxsby replied:
이거 원래 다 하는거 아닌가요? ㅎ ㅎ ? ㅎ ?

daminoworld replied:
크 placeholder색까지 변경해주는 디테일 👏👏 덕분에 NSAttributedString에 대해서 알게됐습니다! 궁금해서 찾아본건데 placeholder 색상도 인스펙터 창에서 변경 가능하더라구요! 이미 수빈선배는 이미 알수도 있지만 혹시나 해서 링크 남깁니다 [링크](https://stackoverflow.com/questions/26076054/changing-placeholder-text-color-with-swift)

lsj8706 commented on Instagram-Clone/Instagram-Clone/Source/Screen/Auth/Controller/LoginViewController.swift:
```diff
@@ -56,27 +61,38 @@ class LoginViewController:
         if (nameTextField.hasText && passwordTextField.hasText) {
             loginButton.isEnabled = true
-            loginButton.backgroundColor = UIColor(displayP3Red: 55/255, green: 151/255, blue: 239/255, alpha: 1)
+            loginButton.backgroundColor = UIColor(named: "sky_blue")
```
색상까지 color assets으로 추가해 주신 건가요? 
디테일 본받겠습니다.

devxsby replied:
이거 편하고 좋더라고요.
다음에 참고하셔서 리팩토링하는 것을 강력추천합니다~~ 😁

daminoworld replied:
와 이거 진짜 좋은것 같아요 직접 코드로 컬러 커스텀 하려니까 너무 지저분 해져서 별로 였는데 좋은 꿀팁 얻어갑니다!

lsj8706 commented on Instagram-Clone/Instagram-Clone/Global/Extension/UITabBarController+.swift:
```diff
@@ -0,0 +1,25 @@
+
+extension UITabBarController {
+    /// 릴스 탭 바 클릭 시 아이템 색상과 에셋 색상을 반전시키는 함수입니다
+    func tabBarReverseCustom() {
```
이 함수의 재사용성을 더 높이기 위해서는 파라미터로 index를 받아와서 해당 인덱스의 탭 바의 속성을 reverse하는것이 어떨까요..?
```suggestion
    func tabBarReverseCustom(index: Int) {
    if selectedIndex == index { ... }
    
```
이런 느낌으로,,,

devxsby replied:
감사합니다 교수님 👑 👑 👑
갓 블레스 유

lsj8706 reviewed (approved):
수고하셨습니다..!
이제 심화 도전 고고

devxsby committed:
[Refactor] tabBarReverseCustom() 리팩토링 (#3)

세진 교수님의 조언을 받아 릴스 탭 바 아이콘 클릭 시 반전되는 함수를 구현하는 부분을 리팩토링하였습니다.

devxsby committed:
[Feat] 누적 뷰 스택 해결을 위한 rootViewController 변경 (#3)

심화 과제 구현입니다.

wngus4296 commented on Instagram-Clone/Instagram-Clone/Global/Extension/UITabBarController+.swift:
```diff
@@ -0,0 +1,25 @@
+            tabBar.tintColor = .black
+            tabBar.unselectedItemTintColor = .black
+        }
+    }
```
저는 여기 `tabBar.barTintColor` 지정해주지않아도 item color 변경이 잘 되던데 넣어주신 이유가 있나요?!?!?

devxsby replied:
탭바는 백그라운드를 원래 `tabBar.barTintColor`로 지정하라고 해서 넣었는데.. 결국 안돼서 `background`로 줬거든요... 🥺
근데 `barTintColor` 지정해도 어차피 적용이 안되니까 지우는게 맞겠네요!
감사합니다!~~!~

wngus4296 commented on Instagram-Clone/Instagram-Clone/Source/Screen/Auth/Controller/AuthCompleteViewController.swift:
```diff
@@ -10,25 +10,45 @@ import UIKit
+        /// rootViewController를 TabBarMainViewController로 변경합니다,
+        guard let rootVC = UIStoryboard(name: "TabBarMain", bundle: nil).instantiateViewController(withIdentifier: "TabBarMainViewController") as? TabBarMainViewController else { return }
+        self.view.window?.rootViewController = rootVC
+        self.view.window?.makeKeyAndVisible()
```
저는 이 코드 안적어줬는데 꼭 필요한 아이인가요? 아니면 있으면 좋은,,?

devxsby replied:
`self.view.window?.makeKeyAndVisible()` 이 코드 말씀하시는 건가요?

32번째 줄 코드에서 이미 핵심적으로 루트 뷰를 바꾸도록 선언을 해줬기 때문에, 33번째 줄 코드는 주현님 말씀대로 굳이 사용하지 않아도 되긴 합니다!!
makeKeyAndVisible()함수는 이제 준비가 되었으니 뷰를 보이게 해라 라는 의미의 함수이고, 일종의 문법처럼 사용하는 것으로 알고 있어요.

저는 PR 포인트에 적은 아래의 이유로 위 makeKeyAndVisible()함수를 사용하였습니다.

---

UIWindow 객체는 가장 최상단에 위치한 화면을 표시하는 객체인데요.  window가 여러개 존재할 때, 가장 앞쪽에 배치된 window를 key window라고 지칭합니다. (보통은 main)

저는 [Apple Developer Documentation](https://developer.apple.com/documentation/uikit/uiwindow)와 [View Programming Guide for iOS](https://developer.apple.com/library/content/documentation/WindowsViews/Conceptual/ViewPG_iPhoneOS/CreatingWindows/CreatingWindows.html)를 참고하여 window의 rootViewController를 위에서 세팅해주고 makeKeyAndVisible() 부르면 마침내 지정한 rootViewController가 상호작용을 받는 현재 화면으로 세팅했습니다~! 

여기 블로그에 해당 설명이 상세하게 적혀있어서 한 번 봐주시면 좋을 것 같아요.! 😇
https://wnstkdyu.github.io/2017/12/29/uiwindow/

88yhtserof replied:
makeKeyAndVisible 함수가 그런 의미였군요!! 저도 공부해보고 코드에 반영해야겠어요!!

wngus4296 commented on Instagram-Clone/Instagram-Clone/Source/Screen/Auth/Controller/AuthCompleteViewController.swift:
```diff
@@ -10,25 +10,45 @@ import UIKit
+    // MARK: - IBAction Part
     @IBAction func completeBtnDidTap(_ sender: Any) {
+        /// rootViewController를 TabBarMainViewController로 변경합니다,
+        guard let rootVC = UIStoryboard(name: "TabBarMain", bundle: nil).instantiateViewController(withIdentifier: "TabBarMainViewController") as? TabBarMainViewController else { return }
```
오 이렇게 할 수 있군여

devxsby replied:
에이 주현님 다 아시면서 ㅎ ㅎ ㅎ, ~~

wngus4296 reviewed (commented):
저도 얼른 리팩토링 갈겨야겠네요 🛠 고생하셨어요!!!!

hyesuuou commented on Instagram-Clone/Instagram-Clone/Global/Extension/UITabBarController+.swift:
```diff
@@ -0,0 +1,25 @@
+
+extension UITabBarController {
+    /// 릴스 탭 바 클릭 시 아이템 색상과 에셋 색상을 반전시키는 함수입니다
+    func tabBarReverseCustom(wantToChangeIndex: Int) {
+        if selectedIndex == 2 {
+            tabBar.backgroundColor = .black
+            tabBar.barTintColor = .white
+            tabBar.tintColor = .white
+            tabBar.unselectedItemTintColor = .white
+        } else {
+            tabBar.backgroundColor = .white
+            tabBar.barTintColor = .black
+            tabBar.tintColor = .black
+            tabBar.unselectedItemTintColor = .black
+        }
```
개인적인 취향이지만 저는 요렇게 같은 속성을 if문으로 나눠줘야 한다면 삼항연산자가 더 눈에 잘 들어와서 삼항연산자를 쓰는 편입니다~!
``` swift
let isReelsSelected = (selectedIndex == 2)
tabBar.backgroundColor = isReelsSelected ? .black : .white
tabBar.barTintColor = isReelsSelected ? .white : .black
// ...
```
요런식으로여!!

devxsby replied:
이 방법 좋은 것 같아요.
이렇게 바꿀게요 ~~ 
감사합니다!! 💫

hyesuuou commented on Instagram-Clone/Instagram-Clone/Source/Screen/Auth/Controller/AuthCompleteViewController.swift:
```diff
@@ -10,25 +10,45 @@ import UIKit
     
-    var message: String?
+    // MARK: - UI Component Part
+    @IBOutlet weak var welcomeMessage: UILabel!
```
welcomeMessageLabel까지 붙이면 더 보기좋은 변수명이 될 것 같아요..!! (이유는 message는 뭔가 그냥 string인 느낌이라서..!!)

devxsby replied:
맞네요.
감사합니다. ㅎㅎㅎ 
역시 완벽한 선배 ❤️

hyesuuou commented on Instagram-Clone/Instagram-Clone/Source/Screen/Auth/Controller/SignUpNameViewController.swift:
```diff
@@ -10,56 +10,58 @@ import UIKit
-            }
+        if nameTextField.hasText {
+            nextButton.isEnabled = true
+            nextButton.backgroundColor = UIColor(named: "sky_blue")
```
여기 색상부분 `UIColor(named: "sky_blue")` 요렇게 써주지 않고도 지금 UIColor라는 extension에 추가하신것 같아서
`.sky_blue`로도 가능할 것 같습니다! (혹은 `UIColor.sky_blue`)

그리고 swift에서 변수명으로 _는 안써서 UIColor extension파일부분에 있는 변수명들 모두 `sky_blue` -> `skyBlue` 요런식으로 카멜케이스로 바꿔주면 좋을 것같아요!!
(요렇게 바꿔주시면 위에 .skyBlue로 해주시면 됩니다!!)

devxsby replied:
아 이거 !!!! 익스텐션으로 지정한거랑 에셋으로 추가한거랑 제가 헷갈려서 일단 통일시켰는데 !!!!💥 
가려웠던 부분 긁어주셔서 넘나 감사합니다,,
컬러 변수 네이밍도 컨벤션 맞출게요~!
혜수 선배는 내거야 ~ ❤️‍🔥

hyesuuou commented on Instagram-Clone/Instagram-Clone/Source/Screen/Auth/Controller/SignUpNameViewController.swift:
```diff
@@ -10,56 +10,58 @@ import UIKit
-                nextButton.isEnabled = false
-                nextButton.backgroundColor = UIColor(displayP3Red: 149/255, green: 200/255, blue: 248/255, alpha: 1)
-            }
+        if nameTextField.hasText {
+            nextButton.isEnabled = true
+            nextButton.backgroundColor = UIColor(named: "sky_blue")
+        }
+        else {
+            nextButton.isEnabled = false
+            nextButton.backgroundColor = UIColor(named: "light_blue")
```
요부분도 개인적인 취향이지만 삼항연산자를 쓰는것도 추천드려여!!

nextButton.isEnabled = nameTextField.hasText
naxtButton.backgroundColor = nameTextField.hasText ? .skyBlue : .lightBlue

devxsby replied:
고쳤습니다 ❤️‍🔥

hyesuuou commented on Instagram-Clone/Instagram-Clone/Source/Screen/Auth/Controller/LoginViewController.swift:
```diff
@@ -10,40 +10,45 @@ import UIKit
+    // MARK: - Life Cycle Part
+    override func viewWillAppear(_ animated: Bool) {
+        super.viewWillAppear(true)
+        deleteAccount()
```
👍 👍

devxsby replied:
칭찬도 해주시고.. 감사합니다 선배 💐

hyesuuou reviewed (commented):
컬러에셋까지 추가해준 디테일 짱.. ❤️  고생하셨습니다~!

hyesuuou commented:
> 로그인 뷰에서 처음에 인스타그램 로고 Image View의 Top을 기준으로 생각하고 피그마에 나온 13mini 기준의 Constrant값을 주고 그 다음 Object들을 종속적으로 연결했습니다.
그러다 보니 화면 사이즈가 작아지면 이미지 뷰가 너무 아래로 내려와서 밑에 Object들이 잘리기 직전까지 나오더라구요.
화면 크기에 맞게 Constrant를 조정하자니 그럼 오토 레이아웃을 한 의미가 없고, 조정을 안하자니 보기가 예쁘지 않은데요.
다들 어떻게 구성하셨는지 궁금해요.

사실 방법이 여러개 있는데, 저는 Top의 Constraint에 현재 핸드폰 화면 크기만큼 비율을 걸어주는 편입니다! 

일단 Constraint도 UIButton이나 UITextField.. 같은 애들이랑 마찬가지로 스토리보드에서선택 -> 끌어서 IBOutlet으로 설정할 수 있어요! 

<img width="431" alt="image" src="https://user-images.githubusercontent.com/68391767/164678415-2b72c39c-ac57-4fa2-9e71-dde7cbc96e0d.png">


그러면 NSLayoutConstraint~~ 라는 타입으로 IBOutlet이 만들어집니다! 이걸 예를들어서 logoTopConstraint라고 이름을 짓는다고 하고 설명해볼게요!!

<img width="529" alt="image" src="https://user-images.githubusercontent.com/68391767/164678476-693d2f0d-85b6-448d-b6fb-ca67004be6a6.png">


현재 핸드폰크기의 실제 height값은 UIScreen.main.bounds.height로 구할 수 있어요!

그래서 지금 상황은 TopConstraint가 126이고, 세로가 812인 폰 기준에서 준거니까

`logoTopConstraint * UIScreen.main.bounds.height * (126/812)`

요런식으로 뷰컨에 적어주면 됩니다!!! 


아니면 지금 보면.. 약간 저 모든 것들이 합쳐서 중간에 있는 느낌이잖아요..?!?! 그래서 저 모든것들을 뷰나 스택뷰에 넣고 가운데 정렬로 갈기는 방법도 있습니다.. 


> 심화 과제 구현을 AppDelegate에서 SceneDelegate를 수정하고 window의 rootViewController 설정하려했는데 keyWindow 등 자꾸 에러가 떠서 쉬운 코드로 변경하였습니다. 이 방식으로 구현하신 분 어떻게 코드 작성하셨는지 궁금해요 !!!
(추가로 이렇게 구현할 경우 MVC패턴이 망가질 수 있다는 글도 읽었는데 왜 그런가요?) 🤔

저도 찾아보니까.. keyWindow라는 속성 자체가 사라진 것 같더라고여..!!!! 그래서 아마 에러가 나는 거 같아요,, 혹시 또 새로운 정보 알고계신분 공유 부탁..

daminoworld commented on Instagram-Clone/Instagram-Clone/Source/Screen/Auth/Controller/LoginViewController.swift:
```diff
@@ -10,40 +10,45 @@ import UIKit
+    }
+    
     override func viewDidLoad() {
         super.viewDidLoad()
-        setupLoginBtn()
+        setUI()
         checkTextField()
```
viewWillAppear에서 deleteAccount()가 먼저 실행되니 checkeTextField()가 실행 되어도 어차피 else 구문으로 실행되어  deleteAccount()만으로 될것 같은데 checkTextField()를 굳이 실행 할 필요가 없지 않을까요..?

devxsby replied:
맞네요 ㅎㅎㅎㅎ. 👍👍👍
제가 deleteAccount()를 나중에 구현했는데 밑에서 중복되는 코드라 삭제하겠습니다~!
감사합니다. 😸

daminoworld commented on Instagram-Clone/Instagram-Clone/Global/Extension/UIViewController+.swift:
```diff
@@ -0,0 +1,18 @@
+
+import UIKit
+
+extension UIViewController {
+    /// Left Back Button Custom
+    func backButtonCustom() {
+        let backBtnIcon = UIImage(named: "icn_back")
+        self.navigationController?.navigationBar.backIndicatorImage = backBtnIcon
+        self.navigationController?.navigationBar.backIndicatorTransitionMaskImage = backBtnIcon
+        self.navigationItem.backButtonTitle = ""
+    }
+}
```
backIndicatorImage랑 backIndicatorTransitionMaskImage 둘 다 설정을 했는데 둘다 이미지를 설정하는것 같은데 어떻게 다른지 궁금해요~!!

devxsby replied:
[Customizing Your App’s Navigation Bar Apple 공식문서](https://developer.apple.com/documentation/uikit/uinavigationcontroller/customizing_your_app_s_navigation_bar)에 `Customize the Back Button with an Image` 부분을 보면 샘플 코드를 아래와 같이 작성했습니다.

```swift
// For Back button customization, setup the custom image for UINavigationBar inside CustomBackButtonNavController.
let backButtonBackgroundImage = UIImage(systemName: "list.bullet")
let barAppearance =
    UINavigationBar.appearance(whenContainedInInstancesOf: [CustomBackButtonNavController.self])
barAppearance.backIndicatorImage = backButtonBackgroundImage
barAppearance.backIndicatorTransitionMaskImage = backButtonBackgroundImage
```
BackButton Image를 커스터마이징할 때
backIndicatorImage 뿐만아니라 말씀하신 [backIndicatorTransitionMaskImage](https://developer.apple.com/documentation/uikit/uinavigationbar/1624938-backindicatortransitionmaskimage)도 같이 설정해주었는데요.

`backIndicatorTransitionMaskImage` 코드를 추가로 작성하는 이유는 push and pop transition(화면 왼쪽 끝 터치해서 밀어넘기는 과정) 중에 content의 mask로 쓰이는 이미지를 추가하기 위함입니다. 

![참고이미지](https://blog.kakaocdn.net/dn/bfleYd/btrcp11KlJb/kWUyJAbEkWrFZ3VDci09m1/img.gif)

추가 설명은 아래 블로그 글 참고해주시면 감사하겠습니다~~! 👇

https://sarunw.com/posts/what-is-backindicatortransitionmaskimage/

88yhtserof replied:
저도 궁금했는데 좋은 정보 알아갑니다!!

ffalswo2 replied:
완전 궁금했던 부분을 긁어주시네요...디테일 장인 수빈님👍 왕초보 스터디때도 공유해주세요!!

daminoworld commented on Instagram-Clone/Instagram-Clone/Source/Screen/Auth/Controller/SignUpNameViewController.swift:
```diff
@@ -10,56 +10,58 @@ import UIKit
+        /// TextField
+        nameTextField.attributedPlaceholder = NSAttributedString(string: "사용자 이름", attributes: [NSAttributedString.Key.foregroundColor : UIColor(named: "dark_grey")!])
+        
+        /// Label
+        nameMessage.text = "새 계정에 사용할 사용자 이름을 선택하세요. 나중에 언제든지 변경할 수 있습니다."
+        self.nameMessage.numberOfLines = 2
```
저도 잘 모르지만 이런 고정적인 메세지는 스토리보드에서 설정하고 코드를 줄이는 것도 좋은것 같아요!

devxsby replied:
이건 제가 스토리보드에서 Label 두 줄로 띄어 쓰면서 오토레이아웃을 헤치지 않게 하는 법을 몰라서 코드로 구현했습니다.
ㅜㅜㅜ

wngus4296 replied:
@devxsby 
<img width="256" alt="image" src="https://user-images.githubusercontent.com/58043306/165499971-db7a1106-6cbb-4d44-82af-c02439928b47.png">

저는 이렇게 해주었는데 요걸 원하시는 게 맞을까요!!!!!?

devxsby replied:
@wngus4296  오 네 맞아요!! 인스펙터에서 하신건가용~~? 
줄 어떻게 띄어쓰신건가요!?

daminoworld commented on Instagram-Clone/Instagram-Clone/Source/Screen/TabBar/Controller/TabBarMainViewController.swift:
```diff
@@ -0,0 +1,35 @@
+        setTabBarColor()
+    }
+
+    // MARK: - Custom Method Part
+    private func setDelegate() {
+      self.delegate = self
+    }
```
viewDidLoad에 바로 self.delegate = self 를 추가하지 않고 함수로 따로 구현한 이유가 있나요?

devxsby replied:
개인적으로 viewDidLoad() 등 앱의 생명주기를 다루는 곳에서는 가독성을 위해 바로 실행 코드를 넣지 않고, 커스텀 함수로 빼놓는 걸 지향하고 있습니다...!

daminoworld commented on Instagram-Clone/Instagram-Clone/Global/Extension/UITabBarController+.swift:
```diff
@@ -0,0 +1,25 @@
+
+import UIKit
+
+extension UITabBarController {
+    /// 릴스 탭 바 클릭 시 아이템 색상과 에셋 색상을 반전시키는 함수입니다
+    func tabBarReverseCustom(wantToChangeIndex: Int) {
+        if selectedIndex == 2 {
+            tabBar.backgroundColor = .black
+            tabBar.barTintColor = .white
+            tabBar.tintColor = .white
+            tabBar.unselectedItemTintColor = .white
+        } else {
+            tabBar.backgroundColor = .white
+            tabBar.barTintColor = .black
+            tabBar.tintColor = .black
+            tabBar.unselectedItemTintColor = .black
+        }
```
tabBarReverseCustom함수의 인자 wantToChangeIndex: Int 가 불필요한 것 같은데 단순 표시 용으로 쓰신건가요?

devxsby replied:
`wantToChangeIndex: Int`는 제 기존 코드에는 없었고, `세진`님의 코드리뷰를 받은 부분인데요.

> 익스텐션에서 이 함수의 재사용성을 더 높이기 위해서는 파라미터로 index를 받아와서 해당 인덱스의 탭 바의 속성을 reverse하는것이 어떨까요..?

담인님 말씀처럼 굳이 필요하진 않지만 TabBarMainViewController()에서도 `tabBarReverseCustom(wantToChangeIndex: 2)` 처럼 호출되기 때문에, 제가 표기용으로 파라미터 값을 볼 수 있어 피드백받은 부분으로 수정하였습니다!

daminoworld commented on Instagram-Clone/Instagram-Clone/Global/Extension/UIViewController+.swift:
```diff
@@ -0,0 +1,18 @@
+
+import UIKit
+
+extension UIViewController {
```
Login, SignUpName, SignUpPassword의 checkTextField()가 다 들어가 있는데 이 함수가 한 개의 텍스트 필드를 검사하는지 두개의 텍스트 필드를 검사하는지만 다르고 구현이 거의 같아서 어떻게 하면 중복을 줄일 수 있을까 고민해봤는데 
여기 extension UIViewController 안에
``` 
func checkTextField(_ firstTextField: UITextField,_ secondTextField: UITextField, _ customButton:UIButton ) {

    if (firstTextField.hasText && secondTextField.hasText) {
            customButton.isEnabled = true
            customButton.backgroundColor = UIColor(named: "sky_blue")
            
        } else {
            customButton.isEnabled = false
            customButton.backgroundColor = UIColor(named: "light_blue")
        }
    }
    
    func checkTextField(_ firstTextField: UITextField, _ customButton:UIButton ) {
        if (firstTextField.hasText) {
            customButton.isEnabled = true
            customButton.backgroundColor = UIColor(named: "sky_blue")

        } else {
            customButton.isEnabled = false
            customButton.backgroundColor = UIColor(named: "light_blue")
        }
    }
```
이렇게 쓰고 각 뷰컨에서 함수 실행시 
``` 
//로그인창
checkTextField(nameTextField, passwordTextField, loginButton)
```
``` 
//가입창(userName)
checkTextField(nameTextField, loginButton)
``` 
이런 식으로하면 각 viewController에서 checkTextField()구문을 빼줄 수 있지 않을까 합니다! (이게 더 좋은지는 모르겠습니다,,ㅎ)

devxsby replied:
말씀해주신대로 구현해보려고 했는데 쉽지않네요 . . .

일단 파라미터를 3개 받는 것(login뷰컨)과 2개 받는 것(sign up뷰컨 2개)을 하나의 함수에서 한 번에 구현하는게 되는건지 조금 더 고민해볼게요.

파라미터를 3개 받는 걸 기본으로 하고 2번째 파라미터를 nil로 하고 되는지 봤는데 이것도 안되네용.. ㅜ 🥺

daminoworld replied:
아 제가 말한거는 파라미터가 2개인거랑 3개인 함수 둘 다 쓰는거였어요!ㅎㅎ 근데 저도 말씀하신것처럼 파라미터가 2개인지 3개인지에 따라 함수 하나로 처리할 수 있게 할 수 있을지도 고민해 봤는데 고부분은 잘 모르겠더군요ㅠ

daminoworld reviewed (commented):
유일한 금잔디 YB 수빈 선배 (주현 선배는 넘 잘해서 YB 아님ㅎㅎ) 감자지만 열심히 코드 리뷰 남겨봤습니다! 이미 너무 깔끔하게 잘해서 제가 제일 열심히 해야할듯ㅠ

devxsby committed:
[Chore] 불필요한 코드 삭제 (#3)

UITabBarController+ 파일에서 background컬러를 적용하는데 사용하지 않는 barTintColor코드를 삭제했습니다.

devxsby committed:
[Chore] TabBar Style translucent 기능 false (#3)

TabBar에서 Style 색상을 투명하게 바꾸는 translucent 기능 false하였습니다.

devxsby committed:
[Style] tabBarReverseCustom 함수 삼항연산자 사용 (#3)

Thanks to 혜수

devxsby committed:
[Style] 로그인 메시지 라벨 네이밍 수정 (#3)

Thanks to 혜수

devxsby committed:
[Fix] 네이밍 수정한 라벨들 스토리보드 재연결 (#3)

devxsby committed:
[Style] Color Asset 코드 네이밍 컨벤션 수정 (#3)

Thanks to 혜수

devxsby committed:
[Style] checkTextField 함수 삼항연산자 사용 (#3)

Thanks to 혜수

devxsby committed:
[Fix] Main 스토리보드 initialViewController 지정 (#3)

devxsby committed:
[Fix] 뒤로가기 버튼 backIndicatorTransitionMaskImage 지정 (#3)

버그 수정

devxsby committed:
[Chore] LoginViewController 중복 코드 삭제 (#3)

LoginViewController에서 viewDidLoad()에서 텍스트 필드 작성상태를 확인하는 checkTextField()함수 호출 코드를 삭제하였습니다.
추가로 deleteAccount()함수의 네이밍을 checkAccount()로 변경하였습니다.

devxsby committed:
[Chore] 텍스트필드 placeholder 텍스트 컬러 지정 스토리보드에서 구현 (#3)

Thanks to 담인
기존 텍스트필드의 placeholder 텍스트 컬러를 바꾸는 기능을 코드에서 구현하던 것을 스토리보드에서 구현하도록 변경했습니다.

devxsby committed:
[Fix] 뒤로가기 버튼 중복 코드 삭제 (#3) 

뒤로가기 버튼을 재사용성을 위해 익스텐션 코드로만 구현하고 Login 스토리보드에서 지정한 이미지 파일을 삭제했습니다.

88yhtserof commented on Instagram-Clone/Instagram-Clone/Global/Extension/UITabBarController+.swift:
```diff
@@ -0,0 +1,18 @@
+
+extension UITabBarController {
+    /// 릴스 탭 바 클릭 시 아이템 색상과 에셋 색상을 반전시키는 함수입니다
+    func tabBarReverseCustom(wantToChangeIndex: Int) {
```
저는 다크모드, 라이트모드를 사용해서 해당 이벤트를 구현했는데 Label과 배경이 설정한 모드에 따라 자동으로 색상이 반전되서 편리하더라고요! 추천합니다~

devxsby replied:
좋은 정보 감사합니다 !!!
근데 하나의 앱에서 라이트 모드와 다크모드를 막 구현하는게 좋은지는 저도 잘 모르겠어요!!
앱이 커지면 유저 반응에 따라 라이트와 다크를 구분해서 구현하고, 컬러 에셋을 라이트 모드와 다크 모드르 따로 지정해두려고 했거든요..!!
그래도 이번 과제에서는 윤휘님이 말씀해주신 걸로 해도 간단해서 좋을 것 같아요~ 굳굳 👍

88yhtserof commented on Instagram-Clone/Instagram-Clone/Global/Extension/UIView+.swift:
```diff
@@ -8,12 +8,5 @@
 import UIKit
 
 extension UIView {
```
빈 파일같은데 남겨두신 이유가 있나요??

devxsby replied:
원래 로그인 완료하고 탭 바로 넘어갈 때 애니메이션 효과를 주려고 익스텐션 UIView+파일에서 [animateWithDuration:animations:completion:](https://developer.apple.com/documentation/uikit/uiview/1622515-animatewithduration) 를 구현하려고 했었는데요.
그냥 구현안해서 코드가 없는 상황입니다 . . .ㅎ... 머쓱타드
다음에 넣도록 할게요 😽

88yhtserof commented on Instagram-Clone/Instagram-Clone/Global/Extension/UIViewController+.swift:
```diff
@@ -0,0 +1,18 @@
+
+extension UIViewController {
+    /// Left Back Button Custom
+    func backButtonCustom() {
```
back 버튼을 커스텀하셨네요!! 생각하지 못했는데 알아갑니다~

다양한 방법을 알면 좋지 않을까해서 제가 사용한 방법도 공유하고 갑니다~~!!
+ 저는 title이 비어있는 버튼 아이템을 만들어서 backButton에 할당해주는 방법을 사용했어요. 이렇게 하면 따로 이미지를 설정해 주지 않아도 돼서 편리했습니다.
 let backButton = UIBarButtonItem()
        backButton.title = ""
        self.navigationController?.navigationBar.topItem?.backBarButtonItem = backButton


+ ++추가로 다른 방법도 남기고 갑니다아아아악
naviagtionController의 왼쪽 back 버튼의 구조가 `< 이전 뷰 컨트롤러 title` 이러한데,
이전 뷰 컨트롤러의 title을 비어있게 설정해줘도 `<` 이렇게 이미지만 남길 수 있어요!!

devxsby replied:
이전에는 뒤로 가기 버튼에 tintColor만 바꿔줬는데, 이번에 asset으로 추가되면서 이미지를 넣어야 하더라구요~
그래서 커스텀 함수를 따로 만들었고, 스토리보드에서 구현하면 다른 네비게이션 컨트롤러에서 사용하지 못하기 때문에 익스텐션에 코드로 따로 구현했습니당 ㅎㅎㅎ
밑에 적어주신 코드는 제가 한거랑 같은것 같네요..!!

88yhtserof commented on Instagram-Clone/Instagram-Clone/Source/Screen/Auth/Controller/LoginViewController.swift:
```diff
@@ -10,73 +10,81 @@ import UIKit
-        }
+        /// 모든 텍스트 필드가 채워져있는 지 확인하는 함수입니다.
+        loginButton.isEnabled = (nameTextField.hasText) && (passwordTextField.hasText)
+        loginButton.backgroundColor = (nameTextField.hasText) && (passwordTextField.hasText) ? UIColor.skyBlue : UIColor.lightBlue
```
저는 버튼의 alpha 값을 사용해 로그인 버튼 색상을 구분해줬어요! 이 방법도 괜찮은 것 같아 추천합니다!

devxsby replied:
alpha 값을 조정하는 것은 디테일이,, 안살아서,,, RGB 커스텀했습니둥 .. ㅎㅎ
개인적으로 RGB 0 0 0 에 alpha값만 조정해서 grey 의 진하기를 나타내는건 좋아보여요~
혹시 모르시면 나중에 함 써보시길 ! 🙋🏻‍♀️

88yhtserof commented on Instagram-Clone/Instagram-Clone/Source/Screen/Auth/Controller/LoginViewController.swift:
```diff
@@ -10,73 +10,81 @@ import UIKit
-        navigationItem.backBarButtonItem?.tintColor = .gray
+    private func checkAccount() {
+        /// Account 작성 상태를 초기화하고 체크하는 함수입니다.
+        nameTextField.attributedText = .none
```
none으로 초기화하는 방법 배워갑니다~~~!!

88yhtserof commented on Instagram-Clone/Instagram-Clone/Source/Screen/TabBar/Model/TabBarMainModel.swift:
```diff
@@ -0,0 +1,8 @@
+//
+//  TabBarMainModel.swift
+//  Instagram-Clone
```
이건 무슨 파일인가요???

devxsby replied:
MVC 패턴 적용할 때 Model은 컨트롤러에서 수행할 내용을 받아 자바 코드, SQL을 수행하고 결과물을 ViewController로 반환하는 역할을 합니다.
저는 아직 쓸 일이 없어서 그냥 파일만 만들어 놓았어요~

88yhtserof commented on README.md:
```diff
@@ -11,7 +11,7 @@
 | 주차 | 내용 | 세미나 | 과제 |
```
README 멋져요,,

devxsby replied:
ㅋㅋㅋㅋㅋ 깨알 칭찬 감사합니다 ❤️‍🔥
쇽삭해서 써먹으시길 😁

yujindonut replied:
저도 쇽샥해가겠습니다 선배 ,, ❤️‍🔥

88yhtserof reviewed (commented):
오늘 처음 수빈님 코드리뷰하는데,,, 역시 최고🙆🏻‍♀️

1번 질문에 대해서는 도움을 드릴 수 있을 것 같아 답변 남겨요!!
+ object가 아래로 내려와 잘리기 직전까지 간다고 하셨는데, bottom에도 constrain을 건 뒤 Priority를 조정해 보는건 어떨까요??
+ Aspect Ratio를 많이 사용하시냐는 물으셨는데,
저는 Aspect Ratio를 자주 사용했어요! 특히 이미지뷰를 화면 별로 사이즈 조정하기 위해 주로 사용했습니다.

제가 인스타그램에 오토레이아웃을 크게 신경쓰지 못했는데, 수빈님처럼 저도 다음 과제 전까지는 완벽하게 만들어야겠어요🔥
자극받고 갑니다~~ 늦어서 죄송해요오오오

devxsby committed:
Merge branch 'main' into assignment2

yujindonut commented:
언니 코드리뷰 진짜 너무 멋있어서 자극받구갑니다❤️‍🔥❤️‍🔥 ,, 이 언니 뭐야뭐야 웅성웅성 멋있기까지 하네 웅성웅성

ffalswo2 commented on Instagram-Clone/Instagram-Clone/Source/Screen/TabBar/Controller/TabBarMainViewController.swift:
```diff
@@ -0,0 +1,35 @@
+}
+
+// MARK: - Extension Part
+extension TabBarMainViewController: UITabBarControllerDelegate {
+    func tabBarController(_ tabBarController: UITabBarController, didSelect viewController: UIViewController) {
+        tabBarReverseCustom(wantToChangeIndex: 2)
+    }
```
index값을 고정값 2로 주어도 가운데 릴스버튼만 눌렸을 때 색상 반전이 잘 되나요??? 어떤식으로 된건지 궁금합니다..!

devxsby replied:
이거 세진이랑 피드백 받아서 고친건데 목요일 왕초보때 설명해드릴게용 !!! 
근데 여기서 넘어가는 파라미터 값은 중요하지 않고(단순 표기용?) tabBarReverseCustom함수에서 확인합니다~!

ffalswo2 replied:
쭉 보고 와서 이해했습니다! 이렇게 넘기는 용도가 아니라 그냥 표시하는 용으로도 써도 상관없나요??? 뭔가 익숙치 않은 방법인데 하나더 알아갑니다

ffalswo2 commented on Instagram-Clone/Instagram-Clone/Source/Screen/Auth/Controller/SignUpPasswordViewController.swift:
```diff
@@ -10,60 +10,60 @@ import UIKit
     
+    // MARK: - IBAction Part
     @IBAction func nextBtnDidTap(_ sender: Any) {
-        guard let authCompleteVC = self.storyboard?.instantiateViewController(withIdentifier: "AuthCompleteViewController") as? AuthCompleteViewController else { return }
-        authCompleteVC.message = messageName
+        /// pop
+        guard let authCompleteVC = UIStoryboard(name: "AuthComplete", bundle: nil).instantiateViewController(withIdentifier: "AuthCompleteViewController") as? AuthCompleteViewController else { return }
```
self.storyboard?가 인식이 똑같이 안되셨나보군요..혹시 원인은 찾으셨나요??

devxsby replied:
### 못찾았는데 그 부분은 민재님이 알아오기로 한 거 아니였나요? ㅎㅎ? ? 예 ? ? ! ?

암튼... 제가 저렇게 바꾼 이유는요~

저는 이전에 Main에서 로그인, 회원가입 2개, 완료 스토리보드를 같이 표현해서 self.로 호출했었는데요.

이번 주차 구현 때 스토리보드를 다 분리하면서 완료 스토리보드를 따로 생성해서 UIStoryboard name을 정확히 지정해주지 않으면 호출이 안돼서 저렇게 적었습니다..!!

self로 호출하는 것 보다 직관적으로 스토리보드 네임이 보여서 저는 저 방식이 더 좋은 것 같더라구요?!

ffalswo2 replied:
네ㅋㅋㅋㅋ맞아요ㅎㅎㅎ그래도 물어봤답니다☺️ 
제가 착각했던 부분은 self가 같은 파일의 경로에 있으면 인식이 된다는 것이였는데 원인이 알고보니 같은 스토리보드 파일내에 있을때만 self.storyboard로 불러올수 있고 다른 스토리보드 파일에 있는 거는 UIStoryboard를 통해서 접근해야되는 거더라고요
수빈님이 말씀해주신 직관적인 부분도 확실히 더 메리트가 있는 것 같아요~!

ffalswo2 reviewed (commented):
디테일 장인 수빈님의 코드와 리뷰하신 글들 보고 많이 배워갑니다👍 수빈님은 왕초보가 아닙니다.
